src/Npgsql/TypeMapping/ConnectorTypeMapper.cs

```diff
@@ -355,55 +356,64 @@ NpgsqlTypeHandler Bind(NpgsqlTypeMapping mapping, PostgresType? pgType = null)
         {
             pgType ??= GetPostgresType(mapping);
             var handler = mapping.TypeHandlerFactory.CreateNonGeneric(pgType, _connector);
-            Bind(handler, pgType, mapping.NpgsqlDbType, mapping.ClrTypes);
-
-            return handler;
+            return GetOrBind(handler, pgType, mapping.NpgsqlDbType, mapping.ClrTypes);
         }
 
-        void Bind(NpgsqlTypeHandler handler, PostgresType pgType, NpgsqlDbType? npgsqlDbType = null, Type[]? clrTypes = null)
+        T GetOrBind<T>(T handler, PostgresType pgType, NpgsqlDbType? npgsqlDbType = null, Type[]? clrTypes = null)
+            where T : NpgsqlTypeHandler
         {
-            if (_handlersByOID.TryGetValue(pgType.OID, out var existingHandler))
+            if (!TryBind(handler, pgType, out var existingHandler, npgsqlDbType, clrTypes))
             {
                 if (handler.GetType() != existingHandler.GetType())
                 {
                     throw new InvalidOperationException($"Two type handlers registered on same type OID '{pgType.OID}': " +
                                                         $"{existingHandler.GetType().Name} and {handler.GetType().Name}");
                 }
 
-                return;
+                // We can cast safely as we just compared a subtype of T (handler.GetType()) against existingHandler.GetType().
+                return (T)existingHandler;
             }
 
-            _handlersByOID[pgType.OID] = handler;
-            _handlersByTypeName[pgType.FullName] = handler;
-            _handlersByTypeName[pgType.Name] = handler;
+            return handler;
 
-            if (npgsqlDbType.HasValue)
+            bool TryBind(NpgsqlTypeHandler handler, PostgresType pgType, [NotNullWhen(false)]out NpgsqlTypeHandler? existingHandler, NpgsqlDbType? npgsqlDbType = null, Type[]? clrTypes = null)
             {
-                var value = npgsqlDbType.Value;
-                if (_handlersByNpgsqlDbType.ContainsKey(npgsqlDbType.Value))
+                var returnedHandler = _handlersByOID.GetOrAdd(pgType.OID, handler);
+                if (!ReferenceEquals(handler, returnedHandler))
```

Comment:
That's the problem I was alluding to - with the previous implementation it was possible for different handlers being registered in `_handlersByOID` and `_handlersByTypeName` due to a race condition in `Bind`. But it looks like @NinoFloris did handle it here.